From 460ec2cadc57c2baaf00ec7055b57eea93555180 Mon Sep 17 00:00:00 2001
From: Natalie Vock <natalie.vock@gmx.de>
Date: Fri, 26 Jun 2026 14:20:10 +0200
Subject: [PATCH 108/114] [FROM-ML] drm/ttm: Add fast-path to evict from
 current bulk move

If we allow evicting our own resources, try looking for resources that
have a lower priority than ours and evict those first.

Signed-off-by: Natalie Vock <natalie.vock@gmx.de>
---
 drivers/gpu/drm/ttm/ttm_bo.c | 32 +++++++++++++++++++++++++++++---
 include/drm/ttm/ttm_bo.h     |  1 +
 2 files changed, 30 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/ttm/ttm_bo.c b/drivers/gpu/drm/ttm/ttm_bo.c
index 295708db7ffa..12a16d313cb6 100644
--- a/drivers/gpu/drm/ttm/ttm_bo.c
+++ b/drivers/gpu/drm/ttm/ttm_bo.c
@@ -562,8 +562,7 @@ static int ttm_bo_alloc_at_place(struct ttm_buffer_object *bo,
 		alloc_state->may_try_low = may_evict;
 
 		may_evict |=
-			dmem_cgroup_below_low(NULL, alloc_state->charge_pool) &&
-			!ctx->cgroup_throttle;
+			dmem_cgroup_below_low(NULL, alloc_state->charge_pool);
 	}
 
 	ret = ttm_resource_alloc(bo, place, res, alloc_state->charge_pool);
@@ -601,6 +600,9 @@ struct ttm_bo_evict_walk {
 	/** @hit_low: If we cannot evict a bo when @try_low is false (first pass) */
 	bool hit_low;
 
+	/** @from_bulk: True if we're evicting from the same ordered bulk move. */
+	bool from_bulk;
+
 	/** @alloc_state: State associated with the allocation attempt. */
 	struct ttm_bo_alloc_state *alloc_state;
 };
@@ -621,7 +623,8 @@ static s64 ttm_bo_evict_cb(struct ttm_lru_walk *walk, struct ttm_buffer_object *
 	 * cgroup is always allowed to evict from itself even if it is protected.
 	 */
 	if (!evict_walk->alloc_state->may_try_low &&
-			bo->resource->css == evict_walk->alloc_state->charge_pool)
+	    bo->resource->css == evict_walk->alloc_state->charge_pool &&
+	    !evict_walk->from_bulk)
 		return 0;
 
 	limit_pool = evict_walk->alloc_state->limit_pool;
@@ -647,6 +650,12 @@ static s64 ttm_bo_evict_cb(struct ttm_lru_walk *walk, struct ttm_buffer_object *
 	evict_valuable = dmem_cgroup_state_evict_valuable(limit_pool, bo->resource->css,
 							  evict_walk->try_low,
 							  &evict_walk->hit_low);
+	/* FIXME: Buffers from the same bulk move always belong to the same VM, but the charged
+	 * owner may errnoneously change when the buffer is evicted. That should be fixed instead
+	 * of having this condition here!
+	 */
+	evict_valuable |= evict_walk->from_bulk;
+
 	if (ancestor)
 		dmem_cgroup_pool_state_put(ancestor);
 
@@ -702,6 +711,23 @@ static int ttm_bo_evict_alloc(struct ttm_device *bdev,
 
 	state->in_evict = true;
 
+	if (ctx->allow_bulk_evict && evictor->bulk_move &&
+	    evictor->bulk_move_order != U32_MAX) {
+		evict_walk.from_bulk = true;
+		lret = ttm_lru_walk_ordered_bulk_for_evict(&evict_walk.walk,
+							   bdev, man,
+							   place->mem_type,
+							   evictor, 1);
+		if (lret)
+			goto out;
+		evict_walk.from_bulk = false;
+	}
+
+	if (!state->may_try_low && ctx->cgroup_throttle) {
+		lret = 0;
+		goto out;
+	}
+
 	evict_walk.walk.arg.trylock_only = true;
 	lret = ttm_lru_walk_for_evict(&evict_walk.walk, bdev, man, 1);
 
diff --git a/include/drm/ttm/ttm_bo.h b/include/drm/ttm/ttm_bo.h
index cf3e169deea5..dcf290cf1667 100644
--- a/include/drm/ttm/ttm_bo.h
+++ b/include/drm/ttm/ttm_bo.h
@@ -179,6 +179,7 @@ struct ttm_operation_ctx {
 	 * when multiple BOs share the same reservation object @resv.
 	 */
 	bool allow_res_evict;
+	bool allow_bulk_evict;
 	/** @cgroup_throttle: Avoid claiming protected memory aggressively. */
 	bool cgroup_throttle;
 	/**
-- 
2.55.0

